### DISCUSSION TITLE
GCC Compiler for C++20

### DISCUSSION THREAD

**Discussion by HamzaHajeir** [esp8266/Arduino#8990] [General]
Hi there

It seems the current compiler is GCC 10 is [experimental and incomplete](https://community.platformio.org/t/std-gnu-20-is-shown-as-201709-nodemcuarduino/32514/2) for supporting c++20. 

When is it planned to upgrade the GCC compiler version (perhaps to GCC 12)?

Thanks,
Hamza

**Comment by mcspr** [comment]
Also see https://github.com/esp8266/Arduino/issues/8916

You should be able to switch version tag in our builder and try for yourself first. It would produce a toolchain package that you would be able to install locally
https://github.com/earlephilhower/esp-quick-toolchain/blob/master/Makefile

Gcc 12 would come with a lot of upstream xtensa changes as well, so not sure how this works with the semver policies applied to our subsequent version bump. May be 3.2.0... but, I suppose it might be better to allow tests instead of having it in sort-of limbo.

**Reply by HamzaHajeir** [reply]
Thanks, I understand that.
 
I've seen GCC11.1 within the file, but the documentation doesn't state 11.1 support there.

Any attention goes with updating to 12.2?

Thanks,

**Reply by earlephilhower** [reply]
You'll need to update multiple of the GNU dependencies.  See https://github.com/earlephilhower/pico-quick-toolchain.  That's moved to 12.3 (and built/run 13.2) cleanly.

The patches in the repo may not cleanly apply (or maybe even be necessary) at GCC 12.x, so be prepared to tweak there too.  ABI differences between the binary blobs from Espressif may also bite you.

If you do get it running, please submit a PR to the https://github.com/earlephilhower/esp-quick-toolchain repo!

**Reply by HamzaHajeir** [reply]
Thanks for your call for envolvement, @earlephilhower. 
As I've no prior knowledge or experience with that kind of development, I've some inquiries about the needed approach:
- To my understanding, upgrading would need to insert [those lines](https://github.com/earlephilhower/pico-quick-toolchain/blob/master/Makefile#L73-L126) to the esp-quick-toolchain's Makefile, correct?
- How to check the unneeded patches here?
- Does the arduino core statically link to Espressif blob (with no rebuild)?
- Checking the espressif/gcc, found a relatively [new release (13.2)](https://github.com/espressif/gcc/tags), isn't that better for cloning or not?

**Reply by earlephilhower** [reply]
Given your questions, I think I might suggest to just try that Espressif released GCC instead.

Building a toolchain for an existing GCC under Linux is only a 1-line command followed by a 30 min wait for all 7 host archs to be built, but every GCC or Binutils revision ends up breaking something or other.  That's just the way it goes, and you need to be able to debug the process.

We have many GCC patches because the Xtensa backend for GCC has/had lots of bugs and generated incorrect assembly code in many cases.  If the patches don't apply then you'd need to figure out why, see if the change in the patch was already implemented, and if not then adjust the patch to make it work w/the new GCC.

Gotchas with the Espressif chain may be related to Newlib (printf, memory management, and pgm_read_xxx most likely).  And any ABI issues because, yes, we do use Espressif blobs.  They don't provide source code so we are stuck w/the NONOS-xxxx.o to link to.

**Reply by HamzaHajeir** [reply]
To update:

Just a small progress until hitting [unknown error](https://github.com/earlephilhower/esp-quick-toolchain/issues/51) to me.
